From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 22 Mar 2026 01:39:17 +0800
Subject: [PATCH] Carpet features


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 315c137cb9dbac51a8ac0d63f0ae0127e57c6c9f..2a2d6aff636ad960638b357d224a631c89d23d33 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -26,6 +26,15 @@ public class ConfigManager {
     public static void initConfigs() {
         registerConfig("luminol", builder.of("luminol", "me.earthme.luminol.config.modules"));
         registerConfig("lophine", builder.of("lophine", "fun.bm.lophine.config.modules")); // add lophine global config
+        // Lophine start - add carpet config
+        registerConfig("lophine_carpet", builder.of(
+                new java.io.File("lophine_config"),
+                "lophine_carpet",
+                "lophine_carpet_config.toml",
+                "lophinecarpetconfig",
+                "fun.bm.lophine.carpet.config.modules"
+        ));
+        // Lophine end - add carpet config
         preLoad();
     }
 
