From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 15 Jun 2026 01:33:32 +0800
Subject: [PATCH] TEMP: Temporarily transform config between different config
 system

Should be removed in Minecraft 26.2 Update

diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 2a2d6aff636ad960638b357d224a631c89d23d33..536e8d7df4ce9de1dbe587c1af6b2474a530a165 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -65,6 +65,12 @@ public class ConfigManager {
         CompletableFuture.allOf(futures).join();
         CommandRegister.register(); // register command after config loaded to enable some command didn't depend on config files
         initialized = true;
+
+        // Lophine start - temporarily transform configs
+        reApplyStagedConfigs();
+        saveConfigs();
+        configfiles.values().forEach((config) -> config.reload(true, false));
+        // Lophine end - temporarily transform configs
     }
 
     public static void registerRunnableBeforeFinalLoad(Runnable runnable) {
