From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 14 Jun 2026 18:10:19 +0800
Subject: [PATCH] Luminol Config System


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3fac32b420ce4ae74166eac43d38c51c1fd513cf..5409e2e6aef23dcf2ec2b0c3488a5cfdb8392277 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3032,6 +3032,17 @@ public final class CraftServer implements Server {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
 
+    // Luminol start - Config API
+    @Override
+    public me.earthme.luminol.api.config.LuminolConfigsInstance getLuminolConfigs(String name) {
+        return me.earthme.luminol.config.ConfigManager.getConfigs(name);
+    }
+
+    public me.earthme.luminol.api.config.LuminolConfigBuilder getLuminolConfigBuilder() {
+        return me.earthme.luminol.config.ConfigManager.getBuilder();
+    }
+    // Luminol end - Config API
+
     // Folia start - region TPS API
     /**
      * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
