From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 14 Jun 2026 18:10:19 +0800
Subject: [PATCH] Luminol Config System


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c1f7510917ebe606280c63d6150829fb0562f474..0cc1b5f4722f382ddea70a61dcea63fa0ae4745a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3022,6 +3022,17 @@ public final class CraftServer implements Server {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
 
+    // Luminol start - Config API
+    @Override
+    public me.earthme.luminol.api.config.LuminolConfigsInstance getLuminolConfigs(String name) {
+        return me.earthme.luminol.config.ConfigManager.getConfigs(name);
+    }
+
+    public me.earthme.luminol.api.config.LuminolConfigBuilder getLuminolConfigBuilder() {
+        return me.earthme.luminol.config.ConfigManager.getBuilder();
+    }
+    // Luminol end - Config API
+
     // Folia start - region TPS API
     /**
      * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
