From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 10:49:40 +0800
Subject: [PATCH] Add tick regions api


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index bb882464831eb500239c44e6b4efbf9000cfee8c..03016bbf7431daefcbb2a7e9487eefc7597636cd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2083,4 +2083,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Luminol start - Tick regions api
+    @Override
+    public me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer() {
+        return this.world.regioniser.threadedRegionizerAPI;
+    }
+    // Luminol end
 }
