From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 10:49:40 +0800
Subject: [PATCH] Add tick regions api


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 6307f788693d3b7a7d33f371fc1ec0bd05af9f12..47a76b9c4029a58421df3e4096f36f62d18df7e4 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2081,4 +2081,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Luminol start - Tick regions api
+    @Override
+    public me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer() {
+        return this.world.regioniser.threadedRegionizerAPI;
+    }
+    // Luminol end
 }
