From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Thu, 2 Jul 2026 22:31:15 +0800
Subject: [PATCH] Correct portal logic of some projectile entities


diff --git a/net/minecraft/world/entity/projectile/ShulkerBullet.java b/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 1b80fa0239ec0c23515639f2b58905eade80412d..6b76ec7868f7ee9676cb4c514185681ae694b13b 100644
--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -229,7 +229,7 @@ public class ShulkerBullet extends Projectile {
         this.setPos(this.position().add(movement));
         this.applyEffectsFromBlocks();
         if (this.portalProcess != null && this.portalProcess.isInsidePortalThisTick()) {
-            this.handlePortal();
+            if (this.handlePortal()) return; // Luminol - Correct portal logic of some projectile entities (Should not be ticking anymore after portal logics)
         }
 
         if (hitResult != null && this.isAlive() && hitResult.getType() != HitResult.Type.MISS) {
diff --git a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
index 89a67fa2d9e9939f77b2311f53624c0fb2f2db93..35aee6c15baeb648de7b8dc6e559f1a26636b21b 100644
--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -274,7 +274,7 @@ public abstract class AbstractArrow extends Projectile {
                     .clipIncludingBorder(
                         new ClipContext(originalPosition, originalPosition.add(movement), ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, this)
                     );
-                this.stepMoveAndHit(blockHitResult);
+                if (this.stepMoveAndHit(blockHitResult)) return; // Luminol - Correct portal logic of some projectile entities (Should not be ticking anymore after portal logics)
             } else {
                 this.setPos(originalPosition.add(movement));
                 this.applyEffectsFromBlocks();
@@ -297,7 +297,7 @@ public abstract class AbstractArrow extends Projectile {
         return 0.99F;
     }
 
-    private void stepMoveAndHit(final BlockHitResult blockHitResult) {
+    private boolean stepMoveAndHit(final BlockHitResult blockHitResult) { // Luminol - Correct portal logic of some projectile entities
         while (this.isAlive()) {
             Vec3 initialPosition = this.position();
             ArrayList<EntityHitResult> entitiesHit = new ArrayList<>(this.findHitEntities(initialPosition, blockHitResult.getLocation()));
@@ -307,7 +307,7 @@ public abstract class AbstractArrow extends Projectile {
             this.setPos(nextLocation);
             this.applyEffectsFromBlocks(initialPosition, nextLocation);
             if (this.portalProcess != null && this.portalProcess.isInsidePortalThisTick()) {
-                this.handlePortal();
+                if (this.handlePortal()) return true; // Luminol - Correct portal logic of some projectile entities (Should not be ticking anymore after portal logics)
             }
 
             if (entitiesHit.isEmpty()) {
@@ -325,6 +325,7 @@ public abstract class AbstractArrow extends Projectile {
                 break;
             }
         }
+        return false; // Luminol - Correct portal logic of some projectile entities
     }
 
     private ProjectileDeflection hitTargetsOrDeflectSelf(final Collection<EntityHitResult> entityHitResults) {
