From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:02:43 +0800
Subject: [PATCH] Fix creative player item pick


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 7dd431fb4b09ca6b8f9b98c2336d889a61ea89b8..f5e051095d0dbb0be7fa06c1817e2ff0567dba2e 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -413,7 +413,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
             Item item = itemStack.getItem();
             int orgCount = itemStack.getCount();
             // CraftBukkit start - fire PlayerPickupItemEvent
-            int canHold = player.getInventory().canHold(itemStack);
+            int canHold = player.hasInfiniteMaterials() ? orgCount : player.getInventory().canHold(itemStack); // Luminol - Fix creative item picking
             int remaining = orgCount - canHold;
             boolean flyAtPlayer = false; // Paper
 
