From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:10:30 +0800
Subject: [PATCH] Fix entity portal-teleport speed


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 61d29f8dc1bd4bed07f6bd1fab4921385e0849c4..1d6f0f8f0da0d619bf199038d6bbb9bb0dcf9596 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1499,7 +1499,42 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         foliaProfiler.startTimer(timerId);
         try {
         // Folia end - profiler
+        // Lophine start - Entity portal-teleport speed fix
         if (isActive) { // Paper - EAR 2
+            if (!(entity instanceof Player) && entity.teleportTickType == fun.bm.lophine.enums.EnumTeleportStage.NEED_SYNC) { // Luminol - after portal compensate tick
+                do { // tick losingTicks + 1 times
+                    entity.tick();
+                } while (entity.losingTicks-- > 0);
+                entity.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NORMAL;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else if (!(entity instanceof Player) && entity.teleportTickType == fun.bm.lophine.enums.EnumTeleportStage.STOP_TICKING) { // Luminol - portal teleport only
+                if (fun.bm.lophine.config.modules.fixes.TeleportConfig.enableDelayCompensation && entity.losingTicks < fun.bm.lophine.config.modules.fixes.TeleportConfig.maxDelayCompensationTicks) {
+                    entity.losingTicks++;
+                }
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+                if (entity.teleportTickType == fun.bm.lophine.enums.EnumTeleportStage.SYNC_NOW) { // teleport canceled
+                    while (entity.losingTicks-- > 0) { // tick losingTicks times because we have added to it already
+                        entity.tick();
+                    }
+                    entity.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NORMAL;
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                        return;
+                    }
+                    if (entity.handlePortal()) {
+                        return;
+                    }
+                }
+            } else {
         entity.tick();
         // Folia start - region threading
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
@@ -1510,6 +1545,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // portalled
             return;
         }
+        }
+        // Lophine end - Entity portal-teleport speed fix
         // Folia end - region threading
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1ca9f9edb5f4195af8eeb5da6bf46071701b735b..ef6cc2c3361a4f20fb3a4230d1bd3d34d325265d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -393,6 +393,8 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public fun.bm.lophine.enums.EnumTeleportStage teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NORMAL; // Lophine - Entity portal-teleport speed fix
+    public int losingTicks = 0; // Lophine - Entity portal-teleport speed fix
 
     public void inactiveTick() {
     }
@@ -3687,6 +3689,7 @@ public abstract class Entity
         } else {
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
+                this.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.STOP_TICKING; // Lophine - Entity portal-teleport speed fix
             } else if (!this.portalProcess.isInsidePortalThisTick()) {
                 this.portalProcess.updateEntryPosition(pos.immutable());
                 this.portalProcess.setAsInsidePortalThisTick(true);
@@ -3711,6 +3714,7 @@ public abstract class Entity
                     // Folia end - region threading
                 } else if (this.portalProcess.hasExpired()) {
                     this.portalProcess = null;
+                    this.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.SYNC_NOW; // Lophine - reset tick type when portal expires
                 }
             }
         }
@@ -4690,6 +4694,7 @@ public abstract class Entity
 
     public void postChangeDimension() {
         this.resetStoredPositions();
+        this.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NEED_SYNC; // Lophine - Entity portal-teleport speed fix
     }
 
     protected static enum PortalType {
