From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 1 May 2026 21:52:07 +0800
Subject: [PATCH] Prevent teleprotAsync calls in move events


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 58830b890ac65bbf17dd56da74dae0574b3eee6a..d6b0f75196dbb167bfeb631100b38158445fd62c 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -741,7 +741,9 @@ public class ServerGamePacketListenerImpl
 
                     Location oldTo = to.clone();
                     PlayerMoveEvent event = new PlayerMoveEvent(player, from, to);
+                    this.player.blockTeleportAsync = true; // Luminol - Prevent teleprotAsync calls during move events
                     this.cserver.getPluginManager().callEvent(event);
+                    this.player.blockTeleportAsync = false; // Luminol - Prevent teleprotAsync calls during move events
 
                     // If the event is cancelled we move the player back to their old location.
                     if (event.isCancelled()) {
@@ -1777,7 +1779,9 @@ public class ServerGamePacketListenerImpl
 
                         Location oldTo = to.clone();
                         PlayerMoveEvent event = new PlayerMoveEvent(player, from, to);
+                                        this.player.blockTeleportAsync = true; // Luminol - Prevent teleprotAsync calls during move events
                         this.cserver.getPluginManager().callEvent(event);
+                                        this.player.blockTeleportAsync = false; // Luminol - Prevent teleprotAsync calls during move events
 
                         // If the event is cancelled we move the player back to their old location.
                         if (event.isCancelled()) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index bdb12df4431a6154c008ec41c1172ac2b7739d2b..2c1e2d694d8e39b0179f08cb700d99e5dba5ea9d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4595,12 +4595,25 @@ public abstract class Entity
             teleportTarget.cause(), teleportFlags, teleportComplete
         );
     }
+    // Luminol start - Prevent teleprotAsync calls in move events
+    public boolean blockTeleportAsync = false;
+    // Luminol end
 
     public final boolean teleportAsync(ServerLevel destination, Vec3 pos, Float yaw, Float pitch, Vec3 velocity,
                                        org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause, long teleportFlags,
                                        java.util.function.Consumer<Entity> teleportComplete) {
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot teleport entity async");
 
+        // Luminol start - Prevent teleprotAsync calls in move events
+        if (this.blockTeleportAsync && me.earthme.luminol.config.modules.fixes.PreventIncorrectTeleportAsyncConfig.enabled) {
+            if (me.earthme.luminol.config.modules.fixes.PreventIncorrectTeleportAsyncConfig.throwWhenCaught) {
+                throw new IllegalStateException("Call teleportAsync during move events!");
+            }
+
+            LOGGER.error("Calling teleportAsync during move events!", new Throwable());
+            return false;
+        }
+        // Luminol end
         if (!ServerLevel.isInSpawnableBounds(new BlockPos(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockX(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockY(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockZ(pos)))) {
             return false;
         }
diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index 6c9ec44a46b6f40a2b8b008cc50c0cafe5e909e6..9260aa8a6ce61819c6c4eeaef5624344e72de5e0 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -284,7 +284,9 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
 
         if (this.lastLocation != null && !this.lastLocation.equals(to)) {
             org.bukkit.event.vehicle.VehicleMoveEvent event = new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, this.lastLocation, to);
+            this.blockTeleportAsync = true; // Luminol - Prevent teleprotAsync calls during move events
             event.callEvent();
+            this.blockTeleportAsync = false; // Luminol - Prevent teleprotAsync calls during move events
         }
         this.lastLocation = vehicle.getLocation();
         // CraftBukkit end
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index db16f4802235646d3e20eebf417951391f3193e8..39d162c6085ed3114795b34c807bb67d468f3aaf 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -319,7 +319,9 @@ public abstract class AbstractMinecart extends VehicleEntity {
         new org.bukkit.event.vehicle.VehicleUpdateEvent(vehicle).callEvent();
 
         if (!from.equals(to)) {
+            this.blockTeleportAsync = true; // Luminol - Prevent teleprotAsync calls during move events
             new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, from, to).callEvent();
+            this.blockTeleportAsync = false; // Luminol - Prevent teleprotAsync calls during move events
         }
         // CraftBukkit end
         this.updateFluidInteraction();
