From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 28 Jun 2026 20:57:48 +0800
Subject: [PATCH] Leaves: Disable packet limit


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 1e16e7c068404a413f0edfa579fd9b3c5b38205a..5d7d944e3aaf6a6cc4b7f7368012e3ddb6473d3c 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -269,8 +269,8 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (this.stopReadingPackets) {
                 return;
             }
-            if (this.allPacketCounts != null ||
-                io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass())) {
+            if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable && (this.allPacketCounts != null || // Luminol - Add config to force disable the packet limiter of Paper
+                io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass()))) { // Luminol - Add config to force disable the packet limiter of Paper
                 long time = System.nanoTime();
                 synchronized (PACKET_LIMIT_LOCK) {
                     if (this.allPacketCounts != null) {
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4e6f0377548b4f5f01c3229ebfb9aabd16a75747..94b9474f2ba1b5088514409ded52f676f159588c 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -910,7 +910,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleCustomCommandSuggestions(final ServerboundCommandSuggestionPacket packet) {
         // Paper start - throttle tab completions
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) {
+        if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable &&!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) {
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM);
             return;
         }
@@ -2187,6 +2187,7 @@ public class ServerGamePacketListenerImpl
     private long lastLimitedPacket = -1;
 
     private boolean checkLimit(long timestamp) {
+        if (me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable) return true; // Leaves - disable
         if (!io.papermc.paper.configuration.GlobalConfiguration.get().spamLimiter.incomingPacketThreshold.enabled()) {
             return true;
         }
@@ -2753,6 +2754,8 @@ public class ServerGamePacketListenerImpl
 
     // Spigot start - spam exclusions
     private void detectRateSpam(final TickThrottler throttler, final String message) {
+        if (me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable) return; // Leaves - disable
+        // CraftBukkit start - replaced with thread safe throttle
         if (org.spigotmc.SpigotConfig.enableSpamExclusions) {
             for (String exclude : org.spigotmc.SpigotConfig.spamExclusions) {
                 if (exclude != null && message.startsWith(exclude)) {
@@ -3552,7 +3555,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handlePlaceRecipe(final ServerboundPlaceRecipePacket packet) {
         // Paper start - auto recipe limit
-        if (!org.bukkit.Bukkit.isPrimaryThread()) {
+        if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable && !org.bukkit.Bukkit.isPrimaryThread()) { // Leaves - can disable
             if (!this.recipeSpamPackets.isIncrementAndUnderThreshold()) {
                 this.disconnectAsync(net.minecraft.network.chat.Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause // Paper - add proper async disconnect
                 return;
