From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 8 Jun 2025 01:23:55 +0800
Subject: [PATCH] Rebrand to Lophine


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..273bd3f354281c7ddbe2a3719be063b1d46e5445 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,17 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // Lophine start
+    /**
+     * The brand id for folia
+     */
+    Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    /**
+     * The brand id for Lophine.
+     */
+    Key BRAND_LOPHINE_ID = Key.key("lophinelabs", "lophine");
+    // Lophine end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
