From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 14 Jun 2026 18:10:20 +0800
Subject: [PATCH] Luminol Config System


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a0e0cb66158cc738476465b6e66fe63eda4f6666..b7b9a17b7c5c7001c9ec4328cd05b22be9e7650d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2778,6 +2778,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
+
+    // Luminol start - Luminol Config API
+    me.earthme.luminol.api.config.LuminolConfigsInstance getLuminolConfigs(String name);
+    me.earthme.luminol.api.config.LuminolConfigBuilder getLuminolConfigBuilder();
+    // Luminol end - Luminol Config API
+
     // Folia start - region TPS API
     /**
      * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
