From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 8 Aug 2026 01:25:33 +0800
Subject: [PATCH] KioCG Chunk API


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index d873b78f6935a23ff8f6092b80968267b182cc0f..b4323370859dd4d9e86614484cdb9d2e35c45b40 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -392,4 +392,6 @@ public interface Chunk extends PersistentDataHolder {
          */
         UNLOADED;
     }
+
+    long getChunkHotAvg(); // KioCG
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 190313bc060d5f8440cec7c249ad1d2f1074f5bc..053b4da2f413e806f3e9590cc6470e22ce500a33 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -4084,4 +4084,6 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @return the post-effects
      */
     PlayerPostEffects postEffects();
+
+    long getNearbyChunkHot(); // KioCG
 }
